### PR TITLE
Refactor config service configuration struct

### PR DESCRIPTION
**Pull request by cw9**
When timeout is not positive, a default timeout will be applied, so
unnecessary to validate it here.

@xichen2020 @prateek

### CODE REVIEW

**Comment by prateek** [comment]
I'm not sure I follow - the following test code causes the old Validate() to fail, but the new one to succeed. Is that the intent?

```
negativeMinute := -1 * time.Minute
err := NewOptions().SetInitTimeout(negativeMinute).Validate()
require.NoError(t, err)
```

**Comment by cw9** [comment]
yes that's the intent, in the code where the value of opts.InitTimeout() is used, it will use to valid value when -1 * time.Minute is returned, I could change the InitTimeout() to return the default value when non positive value is set as an alternative.

**Comment by prateek** [comment]
If -1*time.Minute should not be used, I'd expect Validate() to complain about it. Otherwise what's the point of calling Validate()?

**Comment by cw9** [comment]
The reason I'm removing it is to save the user from providing a value in the config yaml for this field if he or she does not really want to override it. If we want to enforce the users to provide a valid value I'm fine with dropping this diff.

With the validation the empty config will fail since it tries to set it to 0, the Validate() checks the store to be nil so we won't panic on using the stores, but the -1 minute can be and has already been handled by the code by using a valid value instead.

**Comment by xichen2020** [comment]
It might be better to check the timeout value in the config and don't set it if the value is not positive when parsing the configs, as opposed to removing the validation check.

**Comment by cw9** [comment]
It also feels kinda weird cause it's like pushing the logic in sd implementation up here. But I don't feel that strongly either way

**Comment by prateek** [comment]
How about nil checking during config parsing? 

i.e. change config.NewOptions to 
```
func (cfg Configuration) NewOptions() Options {
     opts := NewOptions().
		SetZone(cfg.Zone).
		SetEnv(cfg.Env).
		SetService(cfg.Service).
		SetCacheDir(cfg.CacheDir).
		SetClusters(cfg.etcdClusters())
    if timeout := cfg.M3SD.InitTimeout; timeout != nil {
       opts = opts.SetServiceInitTimeout(timeout) 
    }
    return opts
}
```

Would that allow CreateServices to remain un-changed?

**Comment by cw9** [comment]
> How about nil checking during config parsing?

Yeah that sounds good, refactored the config stuff a bit

File: services/client/config.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+// Configuration is the config for service discovery
+type Configuration struct {
+	InitTimeout *time.Duration `yaml:"initTimeout"`
+}
+
+// NewOptions creates an Option
+func (cfg Configuration) NewOptions() Options {
+	opts := NewOptions()
+	if cfg.InitTimeout != nil {
```
Comment (xichen2020) [inline services/client/config.go]: Is it okay if the config explicitly set the timeout to 0?

Comment (cw9) [inline services/client/config.go]: yeah, it will just fail options.Validation()

Comment (xichen2020) [inline services/client/config.go]: In that case why not just make `InitTimeout` a `time.Duration` and don't call `SetInitTimeout` if `cfg.InitTimeout == 0`?

Comment (cw9) [inline services/client/config.go]: yeah sounds good, removed the pointer

**Pull request merged**